codegen/kuberesource/crd.go

```diff
@@ -153,3 +185,85 @@ func CustomResourceDefinition(
 	}
 	return crd
 }
+
+type ErrMap map[string]error
+type CrdNeedsUpgrade struct {
+	CRDName string
+}
+
+func (e *CrdNeedsUpgrade) Error() string {
+	return fmt.Sprintf("CRD %s needs to be upgraded", e.CRDName)
+}
+
+type CrdNotFound struct {
+	CRDName string
+}
+
+func (e *CrdNotFound) Error() string {
+	return fmt.Sprintf("CRD %s not found. Is it unused?", e.CRDName)
+}
+
+func DoCrdsNeedUpgrade(newProdCrdInfo model.CRDMetadata, ourCrds []apiextv1beta1.CustomResourceDefinition) ErrMap {
+	newProducutVersion := newProdCrdInfo.Version
+	crdmap := make(map[string]string)
+	for _, crd := range newProdCrdInfo.CRDS {
+		crdmap[crd.Name] = crd.Hash
+	}
+	ret := ErrMap{}
+	for _, ourCrd := range ourCrds {
+		var err error
+		var needUpgrade bool
+		if hash, ok := crdmap[ourCrd.Name]; ok {
+			needUpgrade, err = DoesCrdNeedUpgrade(newProducutVersion, hash, ourCrd.Annotations)
+		} else {
+			err = &CrdNotFound{CRDName: ourCrd.Name}
+			ret[ourCrd.Name] = err
+		}
+		if err != nil {
+			ret[ourCrd.Name] = err
+		} else if needUpgrade {
+			ret[ourCrd.Name] = &CrdNeedsUpgrade{CRDName: ourCrd.Name}
+		}
```

Comment:
tried to take a stab at making the logical flow a bit more readable

```suggestion
		if hash, ok := crdmap[ourCrd.Name]; !ok {
			ret[ourCrd.Name] = &CrdNotFound{CRDName: ourCrd.Name}
			continue
		} else {
			needUpgrade, err := DoesCrdNeedUpgrade(newProducutVersion, hash, ourCrd.Annotations)

			if err != nil {
				ret[ourCrd.Name] = err
			} else if needUpgrade {
				ret[ourCrd.Name] = &CrdNeedsUpgrade{CRDName: ourCrd.Name}
			}
		}
```